Captive core fast startup

marta-lokhova opened the pull request:
Implementation based on the discussion in #2960 

Resolves #2960 
Resolves #2993

MonsieurNicolas commented on src/main/CommandLine.cpp:
```diff
@@ -281,10 +281,17 @@ maybeSetMetadataOutputStream(Config& cfg, std::string const& stream)
                               uint32_t startAtLedger,
-                              std::string const& startAtHash)
+                              std::string const& startAtHash,
+                              bool persistPartially = false)
```
do not use default value here

MonsieurNicolas commented on src/main/CommandLine.cpp:
```diff
@@ -916,11 +939,26 @@ int
-        return 0;
-    });
+    auto resetParser = [](bool& resetInMemoryState) {
+        return clara::Opt{resetInMemoryState}["--reset-for-in-memory-mode"](
```
should use `inMemoryParser`: it was added so that we use the same option across subcommands

marta-lokhova replied:
The call `new-db --in-memory` is confusing and contradicts your other comment about calling the database "an in-memory database". The flag is called "in-memory", however the command sets a persistent state, that the next commands also depend on. I'd like to be as clear to the users as possible.

graydon replied:
I suggest reusing the term you're using above: minimal. I.e. `stellar-core new-db --minimal-for-in-memory-mode`

graydon replied:
(I was especially concerned by the word "reset" here because this suggests a non-minimal DB can be changed / reset into a minimal one, and I don't _think_ that's possible?)

MonsieurNicolas commented on src/main/CommandLine.cpp:
```diff
@@ -299,14 +306,30 @@ maybeEnableInMemoryLedgerMode(Config& config, bool inMemory,
+        // aggressively so that we only store a few ledgers worth of data
+        config.AUTOMATIC_MAINTENANCE_PERIOD = std::chrono::seconds(30);
+        config.AUTOMATIC_MAINTENANCE_COUNT = 100;
+        config.DISABLE_XDR_FSYNC = false;
```
you should not touch `DISABLE_XDR_FSYNC` here

marta-lokhova replied:
Picked up the latest master, which already addresses this problem but not forcing `DISABLE_XDR_FSYNC`, so I believe this is resolved now.

MonsieurNicolas reviewed (changes requested):
a few suggestions to make the code flow easier to understand

MonsieurNicolas commented:
@graydon may have better insight than me

marta-lokhova commented on src/herder/test/HerderTests.cpp:
```diff
@@ -2552,7 +2552,7 @@ TEST_CASE("slot herder policy", "[herder]")
 
     // start in sync
-    cfg.FORCE_SCP = true;
+    cfg.FORCE_SCP = false;
```
because it was wrong, look at how the node is configured - it was `NODE_IS_VALIDATOR=false`, but `FORCE_SCP=true`

marta-lokhova commented on src/herder/test/HerderTests.cpp:
```diff
@@ -2565,6 +2565,7 @@ TEST_CASE("slot herder policy", "[herder]")
 
     VirtualClock clock;
     Application::pointer app = createTestApplication(clock, cfg);
+    app->start();
```
because it was wrong, we usually start the application in other tests (see the rest of this file).

marta-lokhova commented on src/main/ApplicationUtils.cpp:
```diff
@@ -131,6 +155,63 @@ runWithConfig(Config cfg, optional<CatchupConfiguration> cc)
+        auto checkBucket = [&](std::string bucketStr) {
+            auto bucketHash = hexToBin256(bucketStr);
+            auto bucket = app.getBucketManager().getBucketByHash(bucketHash);
+            if (!bucket || (isZero(bucket->getHash()) && !isZero(bucketHash)))
```
Did you look at the implementation of `getBucketByHash`? It creates an empty bucket pointer if it can't find the bucket, so checking `!bucket` is not enough.

That being said, looking at the last known ledger loading code, we already check for missing buckets there, so I think I don't even need this check.

MonsieurNicolas replied:
> Did you look at the implementation of `getBucketByHash`? It creates an empty bucket pointer if it can't find the bucket, so checking !bucket is not enough.

yes I did, when it doesn't find the file, it does `return std::shared_ptr<Bucket>();` which is `nullptr` not an empty bucket.

other places in the code are only checking for `nullptr`

But yeah maybe we don't care anyways

marta-lokhova replied:
my bad, for some reason I was convinced that function uses `std::make_shared`. Sorry about that. But yeah, I removed this code now.

marta-lokhova commented on src/main/ApplicationImpl.cpp:
```diff
@@ -134,13 +134,24 @@ ApplicationImpl::ApplicationImpl(VirtualClock& clock, Config const& cfg)
+    // Subtle: initialize the bucket manager first before initializing the
+    // database. This is needed as some modes in core (such as in-memory) use a
+    // small database inside the bucket directory.
+    mBucketManager = BucketManager::create(*this);
```
we don't delete the buckets folder in this case, but I agree it's safer to ensure the order of destruction is consistent with initialization order. Done.

graydon commented on src/main/CommandLine.cpp:
```diff
@@ -957,11 +970,26 @@ int
+    auto resetParser = [](bool& resetInMemoryState) {
+        return clara::Opt{resetInMemoryState}["--reset-for-in-memory-mode"](
+            "Reset the special database used only for in-memory mode (see "
+            "--replay-in-memory flag");
```
the `--replay-in-memory` option is itself marked as deprecated; should _reference_ `--in-memory` here.

graydon commented on src/main/ApplicationUtils.cpp:
```diff
@@ -30,75 +32,125 @@
+}
+
+void
+setupMinimalDB(Config const& cfg, uint32_t startAtLedger)
```
I like the term "minimal DB" you're using here -- I would repeat / reuse / surface this term to the user. It is clear and can be extended into the phrase "minimal DB for in-memory mode".

graydon replied:
(Maybe also extend the name of this function so any reader immediately knows it's about in-memory mode: `setupMinimalDBForInMemoryMode`

graydon commented on src/main/CommandLine.cpp:
```diff
@@ -282,11 +282,17 @@ maybeSetMetadataOutputStream(Config& cfg, std::string const& stream)
-                              uint32_t startAtLedger,
-                              std::string const& startAtHash)
+std::string
+partialDBForInMemoryMode(Config const& cfg)
```
rename "partial" here to "minimal" for consistency.

graydon commented on src/main/CommandLine.cpp:
```diff
@@ -282,11 +282,17 @@ maybeSetMetadataOutputStream(Config& cfg, std::string const& stream)
+std::string
+partialDBForInMemoryMode(Config const& cfg)
+{
+    return fmt::format("sqlite3://{}/partial.db", cfg.BUCKET_DIR_PATH);
```
and `partial.db` here to `minimal.db`

graydon commented on src/main/CommandLine.cpp:
```diff
@@ -957,11 +970,26 @@ int
 runNewDB(CommandLineArgs const& args)
 {
     CommandLine::ConfigOption configOption;
+    bool resetInMemoryState = false;
```
rename variable: `minimalForInMemoryMode`

graydon commented on src/main/ApplicationUtils.cpp:
```diff
@@ -130,6 +182,35 @@ runWithConfig(Config cfg, std::optional<CatchupConfiguration> cc)
 }
 
+bool
+rebuildLatestState(Application& app)
```
rename function for specificity: `applyBucketsOfLastClosedLedger`

graydon commented on src/main/ApplicationUtils.cpp:
```diff
@@ -130,6 +182,35 @@ runWithConfig(Config cfg, std::optional<CatchupConfiguration> cc)
+    auto has = app.getLedgerManager().getLastClosedLedgerHAS();
+    auto lclHash =
+        app.getPersistentState().getState(PersistentState::kLastClosedLedger);
+
```
Add a `releaseAssertOrThrow` here that the ledger tables are all empty (using `LedgerTxn::countObjects`). We never want this code to run on a nonempty ledger.

marta-lokhova replied:
this sounds like a useful check, but the problem is that calling `countObjects` is not allowed on non-root LedgerTxn, which is what we use in in-memory mode (that is, the never committing LedgerTxn). There doesn't seem to be any other machinery to verify LedgerTxn's state (we could add some, but I don't think we should at this time). I will add a comment about this though.

MonsieurNicolas replied:
I agree with Marta, I think it's fine as is for now

graydon commented on src/main/ApplicationImpl.cpp:
```diff
@@ -392,6 +392,24 @@ ApplicationImpl::~ApplicationImpl()
 }
 
+void
+ApplicationImpl::resetDBForInMemoryMode()
```
This worries me / reminds me a bit of a "schema migration", or just .. an alternative path for creating a DB (albeit a minimal one) that does not _necessarily_ arrive at the same place we'd arrive if we reinitialized. I gather you're trying to preserve the overlay state here (but nothing else); would it be possible to do that explicitly? That is:

  1. load all the overlay state you want to preserve into a temporary data structure in memory.
  2. reinitialize the database using the normal code path.
  3. reinsert just the overlay state you wanted to preserve (data which we're a bit lax about anyways)

graydon commented on src/main/ApplicationUtils.cpp:
```diff
@@ -30,75 +32,125 @@
-int
-runWithConfig(Config cfg, std::optional<CatchupConfiguration> cc)
+bool
+canRebuildFromBuckets(uint32_t startAtLedger, uint32_t lcl)
```
rename for specificity: `canRestartInMemoryLedgerFromOldBuckets`

graydon commented on src/main/ApplicationUtils.cpp:
```diff
@@ -30,75 +32,125 @@
+void
+setupMinimalDB(Config const& cfg, uint32_t startAtLedger)
+{
+    VirtualClock clock;
```
Add a `releaseAssert` that we're in in-memory mode.

graydon commented on src/main/ApplicationUtils.cpp:
```diff
@@ -30,75 +32,125 @@
+setupApp(Config& cfg, VirtualClock& clock, bool inMemory,
+         uint32_t startAtLedger, std::string const& startAtHash)
+{
+    if (inMemory)
```
separate flag here seems redundant -- isn't this just `cfg.isInMemoryMode()`?

graydon commented on src/bucket/BucketManagerImpl.cpp:
```diff
@@ -107,7 +107,9 @@ BucketManagerImpl::BucketManagerImpl(Application& app)
-    , mDeleteEntireBucketDirInDtor(app.getConfig().isInMemoryMode())
+    , mDeleteEntireBucketDirInDtor(
+          app.getConfig().isInMemoryMode() &&
+          !app.getConfig().MODE_STORES_HISTORY_LEDGERHEADERS)
```
A comment would be nice here explaining why `!MODE_STORES_HISTORY_LEDGERHEADERS` relates to the deletion of entries in the bucketlist; or perhaps a helper method like `bool Config::isInMemoryModeWithoutMinimalDB() const { return isInMemoryMode() && !MODE_STORES_HISTORY_LEDGERHEADERS; }` or something.

MonsieurNicolas replied:
+1 on adding a helper function for clarity

graydon commented on src/main/Config.h:
```diff
@@ -212,6 +212,9 @@ class Config : public std::enable_shared_from_this<Config>
     // fees, and scp history in the database
     bool MODE_STORES_HISTORY;
 
+    // A config parameter that stores ledger headers in the database
```
Extend this comment to include the fact that this exists in order to support a sub-mode of in-memory mode, where `MODE_STORES_HISTORY` is false but `MODE_STORES_HISTORY_LEDGERHEADERS` remains true. That is: that there is no legal configuration where `MODE_STORES_HISTORY_LEDGERHEADERS` is false but `MODE_STORES_HISTORY` is true. Or in yet other words, that only 3 of the 4 states of these 2 flags are legal.

(Also option to discuss here: I know we've gone back and forth on orthogonal flags vs. enumerated modes, but maybe this is as good as any to ask whether to reconsider merging `MODE_USES_IN_MEMORY_LEDGER`, `MODE_STORES_HISTORY` and `MODE_STORES_HISTORY_LEDGERHEADERS` back into a single enum. I'm not sure which of the _8_ possible combinations is legal off the top of my head, which is not a great sign.)

MonsieurNicolas replied:
mmm, I don't know @graydon : we used to have those "uber flags" used in the code base and it was actually very hard to understand which parts of the code had to change when touching things.

What we may want to do in this commit is probably change `MODE_STORES_HISTORY` to something that doesn't overlap with `MODE_STORES_HISTORY_LEDGERHEADERS` (right now it's hard to tell how they relate to each other: are they mutually exclusive?!).

So if we rename `MODE_STORES_HISTORY` to something like `MODE_STORES_HISTORY_MISC` (and introduce `MODE_STORES_HISTORY_LEDGERHEADERS` before) we can make it clear that they complement each other.

We can add helper functions to help check if *any* historical data is enabled (that seems to be the common case), or maybe to enable/disable "persist all historical data" (to flip both at the same time).

marta-lokhova replied:
I like the idea to rename `MODE_STORES_HISTORY`. I ended up implementing `modeStoresAllHistory` and `modeStoresAnyHistory`, and that seems to improve clarity quite a bit.

MonsieurNicolas replied:
looking good!

graydon reviewed (changes requested):
Overall looks promising! Fairly subtle changes though -- I've marked places where I think things could maybe be made a little more obvious for our future selves when we revisit these treacherous paths (as we surely will). Mostly just cosmetic though, I think there's only one structural / logic change around the question of how to most-safely "reset" vs. "reinitialize + reinsert" a minimal DB that can't be reused.

MonsieurNicolas commented on src/main/ApplicationImpl.cpp:
```diff
@@ -439,10 +439,13 @@ ApplicationImpl::validateAndLogConfig()
     if (getHistoryArchiveManager().hasAnyWritableHistoryArchive())
     {
-        if (!mConfig.MODE_STORES_HISTORY)
+        if (!mConfig.MODE_STORES_HISTORY ||
```
see my other comment: what we're looking for here is to check that if publish is enabled we better store all historical data. So a helper function here would help

MonsieurNicolas commented on src/main/CommandHandler.cpp:
```diff
@@ -75,7 +75,8 @@ CommandHandler::CommandHandler(Application& app) : mApp(app)
 
-    if (mApp.getConfig().MODE_STORES_HISTORY)
+    if (mApp.getConfig().MODE_STORES_HISTORY &&
+        mApp.getConfig().MODE_STORES_HISTORY_LEDGERHEADERS)
```
this should be an `||` as `maintenance` related endpoints should work regardless of which historical data is in there

MonsieurNicolas commented on src/main/CommandLine.cpp:
```diff
@@ -311,17 +313,27 @@ maybeEnableInMemoryLedgerMode(Config& config, bool inMemory,
+        // Since this mode stores historical data (needed to restore
+        // ledger state in certain scenarios), set maintenance to run
+        // aggressively so that we only store a few ledgers worth of data
+        config.AUTOMATIC_MAINTENANCE_PERIOD = std::chrono::seconds(30);
```
any reason to only use `100` here? I think it's possible to replay more than 100 ledgers if captive core is started with a large catchup range (we use `1000000` in the catchup code a little later in the code, maybe we should use the same constant?)

marta-lokhova commented:
@graydon @MonsieurNicolas thanks for the feedback! I believe I addressed your comments in several new commits (I haven't squashed the changes yet, so that it's easy to review).

MonsieurNicolas commented on src/main/ApplicationUtils.cpp:
```diff
@@ -130,6 +195,35 @@ runWithConfig(Config cfg, std::optional<CatchupConfiguration> cc)
+
+    // As the local HAS might have merges in progress, let
+    // `prepareForPublish` convert it into a "valid historical HAS".
+    has.prepareForPublish(app);
```
Another "carry over".

Is this `has.prepareForPublish(app);` really needed?
The current ledger state should never depend on merges in progress (when we close a ledger we have already computed the current ledger) - I think the only thing that this call is going to do is potentially cause a stall when restarting because it will perform merges that it doesn't need to do.

marta-lokhova replied:
This is currently needed since ApplyBucketsWork is designed to apply buckets from an untrusted source, so it verifies validity of the HAS. That being said, I think the check could be moved out of ApplyBucketsWork into catchup. I opened https://github.com/stellar/stellar-core/issues/3044 to track this.

MonsieurNicolas commented on src/main/CommandLine.cpp:
```diff
@@ -1068,26 +1100,82 @@ run(CommandLineArgs const& args)
+                        {
+                            LOG_WARNING(DEFAULT_LOG,
+                                        "Using MANUAL_CLOSE and RUN_STANDALONE "
+                                        "together "
```
this string got chopped-formatted into too many bits. Maybe you need to merge it back into a single line and re-run clang-format?

marta-lokhova replied:
Ah yes, you're right, looks much better now.

MonsieurNicolas commented on src/overlay/PeerManager.cpp:
```diff
@@ -570,6 +570,60 @@ PeerManager::dropAll(Database& db)
+{
+    ZoneScoped;
+    std::vector<std::pair<PeerBareAddress, PeerRecord>> result;
+    std::string sql = "SELECT * FROM peers";
```
avoid using `SELECT *` as it's a footgun, just list the column names that you expect

marta-lokhova replied:
good call, updated.

MonsieurNicolas reviewed (changes requested):
Looks very close to something we can merge. Good job. I added a couple suggestions/questions related to potentially bad things that we're carrying over from master

marta-lokhova commented:
I believe this is ready now @MonsieurNicolas, I've squashed the commits as well.

MonsieurNicolas commented:
r+ 7cde80908478dd16ee5d6358b9a31b77feb38bbe

MonsieurNicolas commented:
@latobarita: retry

MonsieurNicolas commented:
@latobarita: retry
